### PR TITLE
(LAACONNNECT-1357) Implement auto linking scenario for unlinked maat application

### PR DESCRIPTION
**Pull request by pezholio**
## What

[Link to story](https://dsdmoj.atlassian.net/browse/LAACONNECT-1357)

Describe what you did and why.

## Checklist

This handles the "happy path" for auto-linking when the MAAT API returns one result for an imported case. If this is the case we:

- Call `CourtApplicationMaatLinkCreator` (which updates the MAAT ID in Common Platform and posts a link message to the SQS MAAT queue)
- Updates the xhbit case record, setting the `status` to `AUTO_LINKED`, `linked_date` to the current timestamp, and `linked_by `to `SYSTEM`.

### CODE REVIEW

File: app/services/maat_api/maat_application_searcher.rb [hunk 1]
```diff
@@ -23,7 +23,9 @@ def initialize(first_name: nil,
     end
 
     def call
-      @connection.presence&.post(URL, @search_request)
+      MaatApi::SearchResponse.new(
```
Comment (leandroalemao) [inline app/services/maat_api/maat_application_searcher.rb]: Hi @pezholio do you think we need to rescue for some exceptions here in case the post request does not work? or Faraday is configured to not raise exceptions anyway so we don't need to worry about catching exceptions here? https://lostisland.github.io/faraday/#/middleware/included/raising-errors

File: app/services/link_xhibit_case.rb [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+  class LinkFailure < StandardError; end
+
+  def call(maat_search_response, xhibit_case)
+    create_maat_link(xhibit_case.case_urn, maat_search_response.maat_id)
+    update_exhibit_case(xhibit_case, maat_search_response.maat_id)
+  end
+
+private
+
+  def create_maat_link(case_urn, maat_id)
```
Comment (leandroalemao) [inline app/services/link_xhibit_case.rb]: Hi @pezholio what do you think about adding `!` to these methods as they can raise exceptions? just for consistency? tks

**Review by leandroalemao — COMMENTED** [review]
Hi @pezholio LGTM .. just left a small comment really.. tksss 👏 👍

File: app/services/maat_api/maat_application_searcher.rb (hunk 1 above)
Comment (pezholio) [reply app/services/maat_api/maat_application_searcher.rb]: I think this is already handled in `ProcessXhibitCases` here https://github.com/ministryofjustice/laa-court-data-adaptor/blob/7c30ad2df74ecf9daa145c973473140f1532eb5e/app/services/process_xhibit_cases.rb#L17

File: app/services/link_xhibit_case.rb (hunk 2 above)
Comment (pezholio) [reply app/services/link_xhibit_case.rb]: Good plan 👍

File: spec/services/link_xhibit_case_spec.rb [hunk 3]
```diff
@@ -0,0 +1,41 @@
+RSpec.describe LinkXhibitCase, type: :service do
+  describe "#call" do
+    let(:maat_response) { instance_double(MaatApi::SearchResponse, maat_id: "6559879") }
+    let(:xhibit_case) do
+      XhibitMigratedCase.create!(
```
Comment (leandroalemao) [inline spec/services/link_xhibit_case_spec.rb]: Hi @pezholio one for another ticket I guess.. should we add factory bot to the project? thanks

File: app/services/link_xhibit_case.rb [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+  def call(maat_search_response, xhibit_case)
+    create_maat_link!(xhibit_case.case_urn, maat_search_response.maat_id)
+    update_exhibit_case!(xhibit_case, maat_search_response.maat_id)
+  end
+
+private
+
+  def create_maat_link!(case_urn, maat_id)
+    CourtApplicationMaatLinkCreator.call(
+      case_urn,
```
Comment (leandroalemao) [inline app/services/link_xhibit_case.rb]: Hi @pezholio think `CourtApplicationMaatLinkCreator` expects the `subject_id` (`defendant_id`) in the params instead?

File: app/services/link_xhibit_case.rb [hunk 5]
```diff
@@ -0,0 +1,27 @@
+class LinkXhibitCase < ApplicationService
+  class LinkFailure < StandardError; end
+
+  def call(maat_search_response, xhibit_case)
+    create_maat_link!(xhibit_case.case_urn, maat_search_response.maat_id)
```
Comment (leandroalemao) [inline app/services/link_xhibit_case.rb]: Hi @pezholio should we wrap these in a `ActiveRecord::Base.transaction` block? in case any of the actions blows up, it then rolls back.

File: spec/services/link_xhibit_case_spec.rb [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+        sent_date: Date.new(2019, 10, 25),
+      )
+    end
+
+    let(:link_case) { described_class.new.call(maat_response, xhibit_case) }
+
+    before do
+      allow(CourtApplicationMaatLinkCreator).to receive(:call)
+    end
+
```
Comment (leandroalemao) [inline spec/services/link_xhibit_case_spec.rb:31]: Hi @pezholio should we add a `failure path` to the specs? A test with a nil/blank `maat_id` maybe? If we add validations to `XhibitMigratedCase` model to make sure we have `maat_id`, `linked_by` and `linked_at`, when the status is in one the linked states maybe?

**Review by leandroalemao — COMMENTED** [review]
Hi @pezholio I've just added a few more comments... tks 👍

File: spec/services/link_xhibit_case_spec.rb (hunk 3 above)
Comment (pezholio) [reply spec/services/link_xhibit_case_spec.rb]: Oh I agree - was surprised to see it not here to be honest. Happy to add it here actually.

Comment (pezholio) [reply spec/services/link_xhibit_case_spec.rb]: Actually, will make it another PR 👍

File: app/services/link_xhibit_case.rb (hunk 4 above)
Comment (pezholio) [reply app/services/link_xhibit_case.rb]: Ah, nice catch!

File: spec/services/link_xhibit_case_spec.rb [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+        defendant_first_name: "Alice",
+        defendant_last_name: "Smith",
+        sent_date: Date.new(2019, 10, 25),
+      )
+    end
+
+    let(:link_case) { described_class.new.call(maat_response, xhibit_case) }
+
+    before do
+      allow(CourtApplicationMaatLinkCreator).to receive(:call)
```
Comment (leandroalemao) [inline spec/services/link_xhibit_case_spec.rb:29]: Hi @pezholio maybe if you do `allow(CourtApplicationMaatLinkCreator).to receive(:call).and_call_original`, the specs would have failed due the wrong param being sent to the `CourtApplicationMaatLinkCreator`

File: spec/services/link_xhibit_case_spec.rb (hunk 3 above)
Comment (leandroalemao) [reply spec/services/link_xhibit_case_spec.rb]: Hi @pezholio yeah another PR for adding factory bot is a good plan. tks

File: app/services/link_xhibit_case.rb (hunk 5 above)
Comment (pezholio) [reply app/services/link_xhibit_case.rb]: Yeah, I did wonder about that, but I was focussing on the happy path for this PR. That said, I think it's worth adding 👍

File: spec/services/link_xhibit_case_spec.rb (hunk 7 above)
Comment (pezholio) [reply spec/services/link_xhibit_case_spec.rb:29]: Yeah, I think I'd end up having to do a bunch of mocking and setup further down the stack to get the tests working that way. I'd rather keep the unit tests nice and focussed here. It would be a good idea to have to more detailed functional tests though. I'll see where those might belong as part of this.

File: app/services/link_xhibit_case.rb (hunk 5 above)
Comment (pezholio) [reply app/services/link_xhibit_case.rb]: Thinking about it, I don't think there's any value here, as there's only one database call here. If the call to `CourtApplicationMaatLinkCreator` blows up, the call to update the xhibit case won't call, so we're safe there. The only minor risk is if the call to update the xhibit case doesn't succeed, but I think if that happens, we'll have bigger problems.

File: spec/services/link_xhibit_case_spec.rb (hunk 3 above)
Comment (pezholio) [reply spec/services/link_xhibit_case_spec.rb]: PR is here https://github.com/ministryofjustice/laa-court-data-adaptor/pull/1593

File: spec/cassettes/maat_api/search_maat_application_multiple_results.yml [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+        Date:
+          - Fri, 31 Jul 2026 15:22:23 GMT
+        Content-Type:
+          - application/json
+        Transfer-Encoding:
+          - chunked
+        Connection:
+          - keep-alive
+        Set-Cookie:
+          - INGRESSCOOKIE=1785511343.432.31305.664596|55b9a069646b5e4eb9ec5bafd08dbaec;
```
Comment (sophiebrown01) [inline spec/cassettes/maat_api/search_maat_application_multiple_results.yml:90]: This is probably fine but just checking...

File: app/services/link_xhibit_case.rb [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+
+  def call(maat_search_response, xhibit_case)
+    create_maat_link!(xhibit_case.defendant_id, maat_search_response.maat_id)
+    update_exhibit_case!(xhibit_case, maat_search_response.maat_id)
+  end
+
+private
+
+  def create_maat_link!(defendant_id, maat_id)
+    CourtApplicationMaatLinkCreator.call(
```
Comment (sophiebrown01) [inline app/services/link_xhibit_case.rb]: We will also need to update this call downstream (possibly passing the arg in here) to specify that the legal aid status is resubmitted through to CDA and updated back in Common Platform. I think this is missing from the task description referring back to it - let me try and find what the expected param is and get back to you on this.

File: app/services/link_xhibit_case.rb [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+      defendant_id,
+      "SYSTEM",
+      maat_id,
+    )
+  end
+
+  def update_exhibit_case!(xhibit_case, maat_id)
```
Comment (sophiebrown01) [inline app/services/link_xhibit_case.rb]: Small typo, should be `xhibit` not `exhibit` 🙂

**Review by sophiebrown01 — COMMENTED** [review]
Looks good Stuart! A few small points and then the wider discussion in our team channel, but that isn't as a result of this PR 🙂

File: spec/cassettes/maat_api/search_maat_application_multiple_results.yml (hunk 8 above)
Comment (pezholio) [reply spec/cassettes/maat_api/search_maat_application_multiple_results.yml:90]: To be honest, I just copied the cassette and changed the data. It looks like the other cassettes have the same headers in place. We could probably filter that info out when we record the cassettes though, as they're not relevant to the response.

File: app/models/maat_api/search_response.rb [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+    def not_found?
+      status == 404
+    end
+
+    def maat_id
+      response["maatId"]
+    end
+
+    def no_existing_link?
+      !is_linked? && !libra_id && !case_urn
```
Comment (alexdesi) [inline app/models/maat_api/search_response.rb:28]: I am looking for clarification, and need to catch up with the logic:
is not enough `!is_linked?`

File: app/models/maat_api/search_response.rb [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+    def libra_id
+      linking_detail["libraId"]
+    end
+
+    def case_urn
+      linking_detail["caseUrn"]
+    end
+
+    def is_linked?
+      response["isLinked"] == true
```
Comment (alexdesi) [inline app/models/maat_api/search_response.rb:44]: just checking, 
Is response["isLinked"] a boolean (or string)?

File: app/models/maat_api/search_response.rb [hunk 13]
```diff
@@ -0,0 +1,51 @@
+module MaatApi
+  class SearchResponse
+    extend Forwardable
+
+    def_delegator :faraday_response, :body
```
Comment (alexdesi) [inline app/models/maat_api/search_response.rb]: @pezholio we usually tend to avoid `def_delegator` and where possible, be more explicit.
Even if it means being more wordy sometimes.

in this case, this line translates to:
```
def body
  faraday_response.body
end
```

which is Greppable, easier to read, no dependencies (Forwardable)

File: app/models/maat_api/search_response.rb [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+
+    def no_existing_link?
+      !is_linked? && !libra_id && !case_urn
+    end
+
+  private
+
+    attr_reader :faraday_response
+
+    def_delegators :faraday_response, :status, :success?
```
Comment (alexdesi) [inline app/models/maat_api/search_response.rb]: and here:
```
private

def status
  faraday_response.status
end

def success?
  faraday_response.success?
end
```

(Double-check if we need them)

File: app/models/maat_api/search_response.rb (hunk 11 above)
Comment (pezholio) [reply app/models/maat_api/search_response.rb:28]: The scenario mentions all three - https://dsdmoj.atlassian.net/wiki/spaces/ACD2/pages/6092227492/DLRM+Implementation+Analysis#Scenario-3%3A-MAAT-record-found%2C-no-existing-link

File: app/models/maat_api/search_response.rb (hunk 12 above)
Comment (pezholio) [reply app/models/maat_api/search_response.rb:44]: The cassettes seem to suggest it comes out as a boolean

File: app/models/maat_api/search_response.rb (hunk 13 above)
Comment (pezholio) [reply app/models/maat_api/search_response.rb]: That sounds sensible 👍

File: app/services/process_xhibit_cases.rb [hunk 15]
```diff
@@ -1,7 +1,4 @@
 class ProcessXhibitCases < ApplicationService
-  NOT_FOUND = 404
-  SUCCESS = 200..299
```
Comment (alexdesi) [inline app/services/process_xhibit_cases.rb:3]: your refactor looks okay, but double check it.
if I am not wrong, when I tested the API,
there were some scenarios that needed this trick.

File: app/services/process_xhibit_cases.rb [hunk 16]
```diff
@@ -30,18 +27,19 @@ def maat_search(xhibit_case)
...
       handle_not_found(xhibit_case)
     else
       record_error(xhibit_case, error: response&.status, message: response&.body)
     end
   end
 
-  def handle_success(_response, _xhibit_case)
-    # TODO: process the maat applications matching the case.
+  def handle_success(response, xhibit_case)
+    LinkXhibitCase.call(response, xhibit_case) if response.no_existing_link?
```
Comment (alexdesi) [inline app/services/process_xhibit_cases.rb]: if it has already been linked is an important to log it,
a silent failure is difficult to debug

Comment (pezholio) [reply app/services/process_xhibit_cases.rb]: We've got some more tickets to handle the unhappy paths, but I'll add some logging for now 👍

File: spec/models/maat_api/search_response_spec.rb [hunk 17]
```diff
@@ -0,0 +1,187 @@
...
+    it "returns true when is_linked is present but not strictly true" do
+      faraday_response = instance_double(Faraday::Response, status: 200, body: [
+        { "isLinked" => "yes", "linkingDetail" => {} },
+      ])
+      response = described_class.new(faraday_response)
+
+      expect(response.no_existing_link?).to be true
+    end
+  end
+end
```
Comment (alexdesi) [inline spec/models/maat_api/search_response_spec.rb:193]: @pezholio these tests are really thorough, and a very fine-grained level, 
which is not actually bad, 
but to make it more readable we should focus on business logic on a real (fixture json) instead of using tons of instance_double.

What do you think?

Comment (alexdesi) [reply spec/models/maat_api/search_response_spec.rb:193]: some of these tests look like we are testing faraday (but the class is a model `SearchResponse` , we should not care too much on the http client)

File: app/services/link_xhibit_case.rb [hunk 18]
```diff
@@ -0,0 +1,27 @@
+class LinkXhibitCase < ApplicationService
+  class LinkFailure < StandardError; end
```
Comment (alexdesi) [inline app/services/link_xhibit_case.rb]: is this LinkFailure error raised somewhere? 🤔

File: app/services/link_xhibit_case.rb [hunk 19]
```diff
@@ -0,0 +1,27 @@
...
+    create_maat_link!(xhibit_case.defendant_id, maat_search_response.maat_id)
+    update_xhibit_case!(xhibit_case, maat_search_response.maat_id)
+  end
+
+private
+
+  def create_maat_link!(defendant_id, maat_id)
+    CourtApplicationMaatLinkCreator.call(
+      defendant_id,
+      "SYSTEM",
```
Comment (alexdesi) [inline app/services/link_xhibit_case.rb]: is this string deliberate? 🤔 
it sounds we should use a constant for this magic string, 
maybe something like 
const SYSTEM_USER = "SYSTEM"
?

File: app/services/link_xhibit_case.rb (hunk 18 above)
Comment (pezholio) [reply app/services/link_xhibit_case.rb]: Ha, whoops! That was left over from a previous refactor

File: app/services/link_xhibit_case.rb (hunk 19 above)
Comment (pezholio) [reply app/services/link_xhibit_case.rb]: Good plan 👍

File: spec/models/maat_api/search_response_spec.rb (hunk 17 above)
Comment (pezholio) [reply spec/models/maat_api/search_response_spec.rb:193]: I think it's probably OK to use a double for this class, and lean on the fixtures for tests that are more high level (such as the process_xhibit_cases_spec). That said, these tests could be a bit more readable. I'll have a crack :+1:

File: app/services/link_xhibit_case.rb (hunk 9 above)
Comment (pezholio) [reply app/services/link_xhibit_case.rb]: Have addressed this now - see the last three commits 👍

File: app/services/process_xhibit_cases.rb [hunk 20]
```diff
@@ -1,20 +1,22 @@
 class ProcessXhibitCases < ApplicationService
   def call
     XhibitMigratedCase.pending.find_each do |xhibit_case|
-      process_case(xhibit_case)
```
Comment (pezholio) [inline app/services/process_xhibit_cases.rb:7]: This class is getting a bit complex now, with a lot of passing around of `response` and `xhibit_case` variables. I wonder it it'd be easier to have a `ProcessXhibitCase` class that accepts a single `xhibit_case` argument, which might be easier to work with?

File: app/services/link_xhibit_case.rb [hunk 21]
```diff
@@ -1,27 +1,43 @@
...
-  def create_maat_link!(defendant_id, maat_id)
-    CourtApplicationMaatLinkCreator.call(
-      defendant_id,
-      "SYSTEM",
-      maat_id,
-    )
-  end
+  def call
+    case xhibit_case.case_type
+    when TRIAL
```
Comment (pezholio) [inline app/services/link_xhibit_case.rb:14]: I like this!

File: app/services/link_xhibit_case.rb [hunk 22]
```diff
@@ -1,27 +1,43 @@
 class LinkXhibitCase < ApplicationService
-  class LinkFailure < StandardError; end
+  TRIAL = "T".freeze
+  SENTENCE = "S".freeze
+  APPEAL = "A".freeze
+  USER_NAME = "SYSTEM".freeze
```
Comment (pezholio) [inline app/services/link_xhibit_case.rb]: I wonder whether `USER_NAME` should be within the user class, ie: `User::SYSTEM_USERNAME`

File: .rubocop.yml [hunk 23]
```diff
@@ -30,3 +30,6 @@ RSpec/VariableName:
 RSpec/EmptyExampleGroup:
   Exclude:
     - 'spec/requests/**/*.rb'
+
+Rails/Delegate:
+  Enabled: false
```
Comment (pezholio) [inline .rubocop.yml:35]: 👍

File: app/models/maat_api/search_response.rb [hunk 24]
```diff
@@ -1,19 +1,19 @@
 module MaatApi
   class SearchResponse
-    def initialize(faraday_response)
-      @faraday_response = faraday_response
+    def initialize(http_response)
```
Comment (pezholio) [inline app/models/maat_api/search_response.rb:3]: This is much better naming!

**Review by pezholio — COMMENTED** [review]
Looking good - just realised I can't approve, but have made a couple of comments regardless 👍

File: app/services/process_xhibit_cases.rb (hunk 20 above)
Comment (alexdesi) [reply app/services/process_xhibit_cases.rb:7]: I see what you mean, 
I think it's still managable, 
let's see if became more complex building other tickets, in case we'll refactor.
Also I like I can read and reason on a single place instead of splitting it into many mini 3 lines methods.

File: app/services/link_xhibit_case.rb (hunk 22 above)
Comment (alexdesi) [reply app/services/link_xhibit_case.rb]: yes, I agree

File: app/models/user.rb [hunk 25]
```diff
@@ -1,6 +1,8 @@
 # frozen_string_literal: true
 
 class User < ApplicationRecord
+  SYSTEM_USERNAME = "SYSTEM"
```
Comment (pezholio) [inline app/models/user.rb:4]: 👍

File: spec/services/xhibit_auto_linking_spec.rb [hunk 26]
```diff
@@ -0,0 +1,141 @@
...
+      expect(xhibit_case.reload).to be_auto_linked
+      expect(xhibit_case).to have_attributes(
+        maat_id: maat_id.to_s,
+        linked_by: User::SYSTEM_USERNAME,
+        linked_at: within(1.minute).of(Time.zone.now),
+      )
+    end
+
+    it "publishes one MAAT link message with LAA status reinstatement enabled" do
+      expect(Sqs::MessagePublisher).to receive(:call).once.with(
```
Comment (pezholio) [inline spec/services/xhibit_auto_linking_spec.rb]: I tend to prefer using `have_received` after the code under test has been called, particularly as `Sqs::MessagePublisher#call` has already been stubbed. It just reads better to my eyes

Comment (alexdesi) [reply spec/services/xhibit_auto_linking_spec.rb]: yes, it makes sense!

**Review by leandroalemao — APPROVED** [review]
